### DISCUSSION TITLE
Can't login using custom domain

### DISCUSSION THREAD

**Discussion by kcadorin-alpha** [aws-samples/iam-identity-center-team#271] [Q&A] [answered]
I'm deploying directly in the main account so we can control also the main account permissions.

In my first try I deployed without custom domain and everything just worked. The second time I redeployed using a custom domain to make it production ready, but now I can't login on TEAM.

I can't find logs enough to help me understand what is going on.

Seeing a bunch of 304 errors in my custom domain Amplify logs and a bunch of 404 and 307 in the Amplify domain logs.

Attached is the error I get when trying to use my AWS creds to access it.

![image (1)](https://github.com/aws-samples/iam-identity-center-team/assets/171147523/8809bdf2-bfd4-46d2-8b2c-585f74f4eacb)

Help here would be much appreciated.

**Comment by carlaabanes** [comment]
Any progress here? Im about to change our TEAM IDC to use a custom domain and worry about the redirect issues.

**Comment by carlaabanes** [comment, marked as answer]
@kcadorin-alpha did you mean to use the update.sh instructions or did you mean to redeploy everything? I about to do the update.sh with a new custom domain but seeing your topic above makes me worried a bit.

**Reply by kcadorin-alpha** [reply]
I redeployed everything without custom domain by that time but now I had to redeploy in a different account and gave Custom DNS a shot again and found another issue, a bug in the `deployment/integration.sh` and `deployment/cognito.sh` scripts.

Here's the problem and how to fix.

## Problem Description

When using a custom domain with AWS Amplify for the TEAM application, the `integration.sh` and `cognito.sh` scripts incorrectly generate URLs with a "null" prefix (e.g., `https://null.contoso-team.example.com/`), causing SAML authentication failures.

## Root Cause

The issue occurs in the domain detection logic in both `integration.sh` and `cognito.sh` scripts. When a custom domain is configured without a subdomain prefix, the `jq` query returns `null` as a string rather than an empty value:

```bash
# Problematic code (lines 41-42 in integration.sh and lines 37-38 in cognito.sh)
amplifyCustomDomainPrefix=$(echo $amplifyCustomDomains | jq -r 'select(.domainAssociations | length > 0) | .domainAssociations[0].subDomains[] | select(.subDomainSetting.branchName=="main") | .subDomainSetting.prefix')
amplifyDomain=$([ -z "$amplifyCustomDomainPrefix" ] && echo $amplifyCustomDomain || echo $amplifyCustomDomainPrefix.$amplifyCustomDomain)
```

**What happens:**
1. When no `prefix` field exists in the Amplify domain configuration, `jq` returns the string `"null"`
2. The shell script treats `"null"` as a non-empty string
3. The domain is constructed as `null.your-domain.com` instead of `your-domain.com`

## AWS Amplify Domain Configuration Context

For a custom domain like `contoso-team.example.com` configured in Amplify, the JSON structure looks like this:

```json
{
  "domainAssociations": [
    {
      "domainName": "contoso-team.example.com",
      "subDomains": [
        {
          "subDomainSetting": {
            "branchName": "main"
            // Note: No "prefix" field when using root domain
          }
        }
      ]
    }
  ]
}
```

## Impact

- Cognito User Pool Client gets configured with incorrect callback URLs
- SAML authentication redirects to non-existent domains
- Users see "An error was encountered with the requested page" when accessing the application
- The issue affects both the `applicationStartURL` in SAML configuration and Cognito callback URLs

## Solution

Update both `integration.sh` and `cognito.sh` scripts to properly handle missing `prefix` fields:

**In `integration.sh` (lines 40-46):**
```bash
if [ -n "$amplifyCustomDomain" ]; then
  amplifyCustomDomainPrefix=$(echo $amplifyCustomDomains | jq -r 'select(.domainAssociations | length > 0) | .domainAssociations[0].subDomains[] | select(.subDomainSetting.branchName=="main") | .subDomainSetting.prefix // empty')
  if [ -z "$amplifyCustomDomainPrefix" ] || [ "$amplifyCustomDomainPrefix" = "null" ]; then
    amplifyDomain=$amplifyCustomDomain
  else
    amplifyDomain=$amplifyCustomDomainPrefix.$amplifyCustomDomain
  fi
fi
```

**In `cognito.sh` (lines 36-43):**
```bash
if [ -n "$amplifyCustomDomain" ]; then
  amplifyCustomDomainPrefix=$(echo $amplifyCustomDomains | jq -r 'select(.domainAssociations | length > 0) | .domainAssociations[0].subDomains[] | select(.subDomainSetting.branchName=="main") | .subDomainSetting.prefix // empty')
  if [ -z "$amplifyCustomDomainPrefix" ] || [ "$amplifyCustomDomainPrefix" = "null" ]; then
    amplifyDomain=$amplifyCustomDomain
  else
    amplifyDomain=$amplifyCustomDomainPrefix.$amplifyCustomDomain
  fi
fi
```

## Key Changes Explained

1. **`// empty` in jq query**: Returns empty string instead of `null` when field doesn't exist
2. **Explicit null check**: `[ "$amplifyCustomDomainPrefix" = "null" ]` handles cases where `jq` still returns "null"
3. **Cleaner conditional logic**: Uses proper if/else instead of inline conditional

## How to Apply the Fix

1. Edit both `deployment/integration.sh` and `deployment/cognito.sh` files with the corrected logic
2. Re-run the cognito configuration: `./cognito.sh`
3. Verify the output shows correct URLs without "null" prefix

## Verification

After applying the fix, the scripts should output:
```bash
# Before (broken)
applicationStartURL: https://...&redirect_uri=https://null.contoso-team.example.com/&...

# After (fixed)
applicationStartURL: https://...&redirect_uri=https://contoso-team.example.com/&...
```

## Environment Details

- **Affected files**: `deployment/integration.sh`, `deployment/cognito.sh`
- **AWS services**: Amplify (custom domains), Cognito User Pools, IAM Identity Center
- **Domain configuration**: Custom domains without subdomain prefixes
- **Shell**: bash

This issue likely affects any deployment using custom domains without subdomain prefixes in AWS Amplify.

**Comment by carlaabanes** [comment]
Seems to be fixed for our case when I upgraded to v1.4, Thanks!
